# statistical support in phylogenetics


These are the branch support metrics which we are going to explorte:

- Parametric & Nonparametric bootstrap + TBE (trasfer bootstrap expectations)
- Parametric & Nonparametric jacknife
- SH-like approximate likelihood ratio test
- Concordance Factors (Genes & Sites)
- Posterior Probabilities

We can start by inferring a tree specifying a model to skip the model selection step:

```bash
iqtree -s data/example_alignements/nt/Rhabdomeric1.nt.aln -m GTR 
--prefix data/example_alignements/nt/Rhabdomeric1.nt
```

Among the many files generated, there is also a tree: ```data/example_alignements/nt/Rhabdomeric1.nt.nwk.treefile```.

 ---

### nonparametric bootstrap:

The standard nonparametric bootstrap is invoked by the ```-b``` option, which also specifies the number of bootstrap replicates - 100 is the minimum recommended number. We can use the ```--tree-fix``` to fix the toplogy and the ```-t``` flags to specify the tree we just inferred. Otherwise, a new tree will be inferred.

```bash
iqtree -s data/example_alignements/nt/Rhabdomeric1.nt.aln -m GTR -b 100 
--prefix data/example_alignements/nt/parametric_bootstrap 
-t data/example_alignements/nt/Rhabdomeric1.nt.nwk --tree-fix
```

Then explore the result in the ```.iqtree``` file.

---

### parametric bootstrap:

iqtree implements UFB2 (Ultra Fast Bootstrap 2;  described in [Hoang et al., 2018](https://academic.oup.com/mbe/article/35/2/518/4565479)). The ```-B flag``` specifies the number of replicates - 1000 is the minimum number recommended. Perform the parametric bootstrap support using:

```bash
iqtree -s data/example_alignements/nt/Rhabdomeric1.nt.aln -m GTR -bb 1000 
--prefix data/example_alignements/nt/parametric_bootstrap
```

__NB:__ due to the functioning of the parametric bootstrap a tree inference has to be performed and it is not bossible to 

---

### Question:

> ```parametric_bootstrap.splits.nex```

Among the large number of parameters which can affect the tree search process in IQ-TREE, IMHO some of the more decisive are:

Bootstrap options:

 * ```-bnni``` = further optimitation of each bootstrap tree using a hill-climbing nearest neighbor interchange (NNI) search, based directly on the corresponding bootstrap alignment. Usefull to reduce the risk of overestimating branch supports with UFBoot due to severe model violations. 

 * ```-nm``` = Specify the maximum number of iterations to stop (*DEFAULT* 1000). If after 1000 iterations bootstrap values still not convergence, it will give up (usually it is increased).

### Question:

> try to repeat the analyses with different parameters ... do you see results changing?

--- 

### Jacknife:




```bash
iqtree -s data/example_alignements/nt/Rhabdomeric1.nt.aln -m GTR -bb 1000 
--prefix data/example_alignements/nt/parametric_bootstrap
```


--- 

### SH-like approximate likelihood ratio test:

IQ-TREE implements a non-parametric approximate likelihood ratio test based on a Shimodaira-Hasegawa-like procedure _via_ the flag ```-alrt```.

```bash
iqtree -s data/example_alignements/nt/Rhabdomeric1.nt.aln -m GTR -alrt 1000 --prefix data/example_alignements/nt/alrt
```

You can also perform both SH-aLRT and the ultrafast bootstrap on the "best" Maximum Likelilhood phylogeny. It's then easier to observe wether the different support metrics are giving contrasting results through our phylogenies.

```bash
iqtree -s data/example_alignements/nt/Rhabdomeric1.nt.aln -m GTR -B 1000 -alrt 1000 --prefix data/example_alignements/nt/cobined_supports
```

Let's take a look at the ```.iqtree``` among the other output files, bearing in mind that the values related to different metrics should be treated differently: with the non-parametric bootstrap and SH-aLRT you should start to _believe_ in a clade if it has >= 80% support, while with UFBoot it should be >= 95%.

--- 

### CFs

([Minh et al., 2020](https://academic.oup.com/mbe/article/37/9/2727/5828940)) two measures for quantifying genealogical concordance in phylogenomic datasets: the gene concordance factor (**gCF**) and the site concordance factor (**sCF**). For every branch of a reference tree:

- gCF is defined as the percentage of “decisive” gene trees containing that branch. 

- sCF is defined as the percentage of decisive alignment sites supporting a branch in the reference tree. 


gCF is already in wide usage, but here we allow to calculate gCF while correctly accounting for variable taxon coverage among the gene trees. sCF is a novel measure that is particularly useful when individual gene alignments are relatively uninformative, such that gene trees are uncertain.

Let's start by ingerring a tree:

```bash
iqtree2 -p ALN_DIR --prefix concat -B 1000 -T AUTO
```


```bash
iqtree2 -S ALN_DIR --prefix loci -T AUTO
```

Now let's perform CFs calculation with the following command:

- for gene CFs

```bash
iqtree2 -t concat.treefile --gcf loci.treefile --prefix concord
```

- for sites CFs
```bash
iqtree2 -t concat.treefile --gcf loci.treefile --prefix concord
```

and take a look at the resulting nodal support values on figtree (compared them each other and against BS).

--- 

## further reading: 

A couple of nice papers on how the nodal support can be overestimated by [BP](https://doi.org/10.1016/j.ympev.2014.01.018) and [PP](https://doi.org/10.1080/10635150590924208). 

Resources on concordance factors: [paper](https://www.biorxiv.org/content/10.1101/487801v2) & [tutorial](http://www.robertlanfear.com/blog/files/concordance_factors.html).

[Here](http://www.iqtree.org/doc/iqtree-doc.pdf) you'll find the manual for IQ-TREE, it's quite user-friendly and exhaustive. Many tutorials are based on those described here.

---

### [main](https://github.com/for-giobbe/MP25/tree/main)

