# statistical support in phylogenetics


These are the branch support metrics which we are going to explorte:

- Parametric & Nonparametric bootstrap + TBE (trasfer bootstrap expectations)
- Parametric & Nonparametric jacknife
- SH-like approximate likelihood ratio test
- Concordance Factors (Genes & Sites)

---

### Nonparametric bootstrap:

The standard nonparametric bootstrap is invoked by the -b option, which also specifies the number of bootstrap replicates (100 is the minimum recommended number).

```bash
 iqtree -s ND2_p_aligned.n.gb.fasta -b 100
```

 --- 

### Parametric bootstrap:

IQ-TREE implements UFB2 - Ultra Fast Bootstrap 2 described in [Hoang et al., 2018](https://academic.oup.com/mbe/article/35/2/518/4565479). The -B flag specifies the number of replicates where 1000 is the minimum number recommended.

```bash
 iqtree -s ND2_p_aligned.n.gb.fasta -bb 1000 
```

Among the large number of parameters which can affect the tree search process in IQ-TREE, IMHO some of the more decisive are:

Bootstrap options:

 * ```-bnni``` = further optimitation of each bootstrap tree using a hill-climbing nearest neighbor interchange (NNI) search, based directly on the corresponding bootstrap alignment. Usefull to reduce the risk of overestimating branch supports with UFBoot due to severe model violations. 
 * ```-nm``` = Specify the maximum number of iterations to stop (*DEFAULT* 1000). If after 1000 iterations bootstrap values still not convergence, it will give up (usually it is increased).

 --- 

### SH-like approximate likelihood ratio test:

IQ-TREE implements a non-parametric approximate likelihood ratio test based on a Shimodaira-Hasegawa-like procedure via the flag -alrt.

 iqtree -s ND2_p_aligned.n.gb.fasta -alrt 1000

--- 

We can combine the three metrics in the same analysis and have them annotated on the "best" Maximum Likelilhood phylogeny. It's then easier to observe wether the different support metrics are giving contrasting results through our phylogenies.

iqtree -s ND2_p_aligned.n.gb.fasta -bb 1000 -bnni -b 10 -alrt 1000 -redo

Let's take a look at the .iqtree among the other output files, bearing in mind that the values related to different metrics should be treated differently: with the non-parametric bootstrap and SH-aLRT you should start to believe in a clade if it has >= 80% support, while with UFBoot it should be >= 95%,

 --- 

### CFs

([Minh et al., 2020](https://academic.oup.com/mbe/article/37/9/2727/5828940)) two measures for quantifying genealogical concordance in phylogenomic datasets: the gene concordance factor (**gCF**) and the site concordance factor (**sCF**). For every branch of a reference tree, gCF is defined as the percentage of “decisive” gene trees containing that branch. sCF is defined as the percentage of decisive alignment sites supporting a branch in the reference tree. sCF is a novel measure that is particularly useful when individual gene alignments are relatively uninformative, such that gene trees are uncertain.

Now let's perform CFs calculation with the following command:

```
mkdir Analyses/IQ-TREE/CFs
iqtree -t Analyses/IQ-TREE/specie.tree.treefile --gcf Analyses/IQ-TREE/Gene_Trees/loci.treefile -p Analyses/1-to-1_Alignments/gblock/ --scf 100 --prefix Analyses/IQ-TREE/CFs/concord
```

and take a look at the resulting nodal support values on figtree (compared them each other and against BS).

























## further reading: 

A couple of nice papers on how the nodal support can be overestimated by [BP](https://doi.org/10.1016/j.ympev.2014.01.018) and [PP](https://doi.org/10.1080/10635150590924208). 

Resources on concordance factors: [paper](https://www.biorxiv.org/content/10.1101/487801v2) & [tutorial](http://www.robertlanfear.com/blog/files/concordance_factors.html).

[Here](http://www.iqtree.org/doc/iqtree-doc.pdf) you'll find the manual for IQ-TREE, it's quite user-friendly and exhaustive.

---

### [main](https://github.com/for-giobbe/MP25/tree/main)

