# statistical support in phylogenetics

These are the branch support metrics which we are going to explorte:

- parametric & nonparametric Bootstrap Proportions (BPs) + trasfer bootstrap expectations (TBE)
- parametric & nonparametric jacknife
- SH-like approximate Likelihood Ratio Test (SHaLRT)
- genes & sites Concordance Factors (gCFs & sCFs)
- Posterior Probabilities (PP)

We can start by inferring a tree specifying a model to skip the model selection step:

```bash
iqtree -s data/example_alignements/nt/Rhabdomeric1.nt.aln -m GTR 
--prefix data/example_alignements/nt/Rhabdomeric1.nt
```

Among the many files generated, there is also a tree: ```data/example_alignements/nt/Rhabdomeric1.nt.nwk.treefile```.

 ---

### nonparametric bootstrap:

The standard nonparametric bootstrap is invoked by the ```-b``` option, which also specifies the number of bootstrap replicates - 100 is the minimum recommended number. We can use the ```--tree-fix``` to fix the toplogy and the ```-t``` flags to specify the tree we just inferred. Otherwise, a new tree will be inferred.

```bash
iqtree -s data/example_alignements/nt/Rhabdomeric1.nt.aln -m GTR -b 100 
--prefix data/example_alignements/nt/nonparamet_bootstrap 
-t data/example_alignements/nt/Rhabdomeric1.nt.nwk --tree-fix
```

Then explore the result in the ```.iqtree``` file.

---

### parametric bootstrap:

iqtree implements UFB2 (Ultra Fast Bootstrap 2;  described in [Hoang et al. 2018](https://academic.oup.com/mbe/article/35/2/518/4565479)). The ```-B flag``` specifies the number of replicates - 1000 is the minimum number recommended. Perform the parametric bootstrap support using:

```bash
iqtree -s data/example_alignements/nt/Rhabdomeric1.nt.aln -m GTR -bb 1000 
--prefix data/example_alignements/nt/parametric_bootstrap
```

__NB:__ due to the functioning of the parametric bootstrap a tree inference has to be performed and it is not bossible to calculate the BPs on a previously inferred tree.

---

### Question:

> Take a look at the  ```parametric_bootstrap.splits.nex``` file ... does it remember anything to you?!

> default resample the sites within partitions however, it is recommended to resample partitions (Nei et al., 2001). This can be done with -bsam GENE option. Moreover, it is also possible to resample partitions and then sites within resampled partitions (Gadagkar et al., 2005). This may help to reduce false positives (i.e. wrong branch receiving 100% support). This can be done with -bsam GENESITE.

---

### TBE:

To implement the transfer bootstrap expectaion in our anlysius is as simple as adding a single flag --tbe in iwtree:


```bash
iqtree -s data/example_alignements/nt/Rhabdomeric1.nt.aln -m GTR -bb 1000 
--prefix data/example_alignements/nt/parametric_bootstrap_TBE --tbe
```

### Question:

Is the resul chanigng with and withou the use of TBE?! Think in which kind of dataset you imagine

--- 

### Jacknife:

The non-replacement version of the boostrap can also be performed in a parametric and non-parametric fashion using the ```-j``` and ```-J``` flags:

```bash
iqtree -s data/example_alignements/nt/Rhabdomeric1.nt.aln -m GTR -j 100 
--prefix data/example_alignements/nt/nonparamet_jacknife
```

and

```bash
iqtree -s data/example_alignements/nt/Rhabdomeric1.nt.aln -m GTR -J 1000 
--prefix data/example_alignements/nt/parametric_jacknife
```

Other than the number of replicate, theubsampling proportion for jackknife cna be specified using the flag ```--jack-prop```.

--- 

### SH-like approximate likelihood ratio test:

Among single branch tests, iqtree implements a non-parametric approximate likelihood ratio test based on a Shimodaira-Hasegawa-like procedure _via_ the flag ```-alrt```.

```bash
iqtree -s data/example_alignements/nt/Rhabdomeric1.nt.aln -m GTR -alrt 1000 --prefix data/example_alignements/nt/alrt
```

You can also perform both SH-aLRT and the ultrafast bootstrap on the "best" Maximum Likelilhood phylogeny. It's then easier to observe wether the different support metrics are giving contrasting results through our phylogenies.

```bash
iqtree -s data/example_alignements/nt/Rhabdomeric1.nt.aln -m GTR -B 1000 -alrt 1000 --prefix data/example_alignements/nt/cobined_supports
```

Let's take a look at the ```.iqtree``` among the other output files, bearing in mind that the values related to different metrics should be treated differently: with the non-parametric bootstrap and SH-aLRT you should start to _believe_ in a clade if it has >= 80% support, while with UFBoot it should be >= 95%.

### Question:

> We are just scratching the surface of ```--abayes```

--- 

### CFs

([Minh et al., 2020](https://academic.oup.com/mbe/article/37/9/2727/5828940)) two measures for quantifying genealogical concordance in phylogenomic datasets: the gene concordance factor (**gCF**) and the site concordance factor (**sCF**). For every branch of a reference tree:

- gCF is defined as the percentage of “decisive” gene trees containing that branch. 

- sCF is defined as the percentage of decisive alignment sites supporting a branch in the reference tree. 

Let's start by inferring a species tree:

```bash
iqtree2 -p data/example_alignements/aa --prefix concat
```

Then we can infer a tree for each locus:

```bash
iqtree2 -S data/example_alignements/aa --prefix loci
```

Now let's perform CFs calculation with the following command:

```bash
iqtree -p data/example_alignements/aa -t concat.treefile --gcf loci.treefile --scf 100 --prefix concord
```
and take a look at the resulting nodal support values on figtree. 

### Question:

> (compared them each other and against BS)

--- 

### posterior proportions

 the posterior probability is calculated as the fraction of sampled trees that include that clade

--- 

## further reading: 

A couple of nice papers on how the nodal support can be overestimated by [BP](https://doi.org/10.1016/j.ympev.2014.01.018) and [PP](https://doi.org/10.1080/10635150590924208). 

Resources on concordance factors: [paper](https://www.biorxiv.org/content/10.1101/487801v2) & [tutorial](http://www.robertlanfear.com/blog/files/concordance_factors.html).

[Here](http://www.iqtree.org/doc/iqtree-doc.pdf) you'll find the manual for IQ-TREE, it's quite user-friendly and exhaustive. Many tutorials are based on those described here.

---

### [main](https://github.com/for-giobbe/MP25/tree/main)

