# systematic and stochastic bias + phylogenetic subsampling

To intiuitively get a grasp of one of the more frquent phylogenetic artifacts - long branch attractio or LBA - open the file ```bias.aln``` in the folder ```bias.aln```. I have duplicated two sequences Ptetracirrhus_bias and Ovulgaris_bias, and I wish you to insert as may As and Ts as you can in these squences, kind of simulating a parallel evolution of these two lineage, but where a lot more evolutionary changes happened, but only proe to GC. ... then infer a quick tree using ```iqtree -s data/example_bias/biased.aln``` and take a look at the tree ...

---

# mixture models

Let's start by concatenating again the MSAs

```bash
AMAS.py concat -i data/example_bias/alignements/* -t bias_concat.faa -f fasta -d aa
```

Then we can perform two inference:

- a first one assuming a simple model with precasted matrix
```bash
iqtree -s bias_concat.faa -mset LG -mrate G4 --prefix bias_simple -nt auto -fast
```

- a second one with C60:

```bash
iqtree -s bias_concat.faa -mset LG+C60 -mrate G4 --prefix bias_mixture -nt auto -fast
```

### task!

> Are the two trees different? It is kind of hard to tell ...
---

# data recoding

recoding using python script

---

# HRS assumptions

Traditional model relyes on the assumption that evolution of DNA sequences follow a markov process. However, again, it may not **always** be true. For this reason IQ-TREE implements three *matched-pairs tests of symmetry* ([Naser-Khdour et al., 2019](https://academic.oup.com/gbe/article/11/12/3341/5571717)) to test the two assumptions of *stationarity* and *homogeneity*. Morever are also implemented some Lie Markov models ([Woodhams et al., 2015](https://academic.oup.com/sysbio/article/64/4/638/1650486)), many of which are non-reversible models.

```bash
iqtree -S data/example_bias/alignments/ --symtest-only
```


The three important columns are:

SymPval: a small p-value (say < 0.05) indicates that the assumptions of stationarity or homogeneity or both is rejected. In this case, partition part3 does not comply with these two assumptions (p-value = 0.00499855), whereas the other two partitions are “good”.

MarPval: a small p-value means that the assumption of stationarity is rejected. In this case, only partition part3 does not comply with the stationary condition (p-value = 0.00164132).

IntPval: a small p-value means that the homogeneity assumption is reject. In this case, no partitions are “bad” according to this test, i.e., they all comply with the homogeneity assumption.


---

LMAP to use only informative genes

```bash
iqtree -s 100%_aa_aln.fa -lmap 2000 -n 0 -m LG
```

---

### [main](https://github.com/for-giobbe/MP25/tree/main)
