### inferring selection

Phylogenetics aims is to reconstruct the patterns of relationships among different organisms. And while phylogenetic trees can be considered by themselves the results of such investigations, they can also be used to study evolutionary processes with a myriad of possibilities.

---

Here we are going to carry out likelihood estimations of the dN/dS ratio [(Goldman and Yang 1994)](https://doi.org/10.1093/oxfordjournals.molbev.a040153) using CODEML, which is part of the PAML package [(Ziheng Yang, 2007)](https://academic.oup.com/mbe/article/24/8/1586/1103731). PAML is an extremely versatile piece of software whose features include:

* estimating synonymous and nonsynonymous rates
* testing hypotheses concerning dN/dS rate ratios
* ancestral sequence reconstruction (DNA, codon, or AAs)
* various clock models
* simulating nucleotide, codon, or AA sequence data sets
* ...

Just an extensive reding of the [manual](http://abacus.gene.ucl.ac.uk/software/pamlDOC.pdf) will disclose all
of its possibilities! However, several great and newer alternative exist, the most notable being HyPhy [(Pond et al., 2019)](https://doi.org/10.1093/molbev/msz197) which is also implemented in the Datamonkey [web server](https://www.datamonkey.org/). 

---

Biologically we will try to answer to the question: do squid species inhabiting the depth of the species have specific adaptations to low-light environiments? In this tutorial we will use a gene encoding a rhabdomeric opsin, possibly one of the primary opsins used in a specific organismâ€™s visual system. Here are the species which we will consider:

- Ovulgaris - [_Octopus vulgaris_](https://upload.wikimedia.org/wikipedia/commons/thumb/e/ed/Polpo_nella_Riserva_naturale_statale_Torre_Guaceto_-_DSC_0788M.jpg/1200px-Polpo_nella_Riserva_naturale_statale_Torre_Guaceto_-_DSC_0788M.jpg)
- Emoschata - [_Eledone moschata_](https://www.monaconatureencyclopedia.com/wp-content/uploads/2008/08/3-Eledone-moschata.jpg)
- Cmacropus - [_Callistoctopus macropus_](https://upload.wikimedia.org/wikipedia/commons/8/82/Callistoctopus_macropus_en_col%C3%A8re_%28photo_de_nuit%29.jpg)
- Ecirrhosa - [_Eledone cirrhosa_](https://upload.wikimedia.org/wikipedia/commons/0/00/Eledone_cirrhosa_Merculiano.jpg) 
- Ptetracitrrhus - [_Pteroctopus tetracirrhus_](https://litoraldegranada.ugr.es/wp-content/uploads/2019/01/PTPORT-960x412.jpg)
- Sunicirrhus - [_Scaeurgus unicirrhus_](https://upload.wikimedia.org/wikipedia/commons/3/33/Scaeurgus_unicirrhus.jpg)
- Aargo - [_Argonauta argo_](https://upload.wikimedia.org/wikipedia/commons/a/a1/Argonauta_argo_Merculiano.jpg)


Among them the eep sea speciesare :
- _Eledone moschata_
- _Scaeurgus unicirrhus_
- _Pteroctopus tetracirrhus_

---

Let's check what we need in order to run CODEML:

* a ```.fasta``` alignment
* a ```.nwk``` tree
* a ```.ctl``` file - where we configure an analysis

We miss the ```.ctl``` - which stads for control file - and has a similar role to the ```.cfg``` we've seen earlier.
Here is how it looks like:

```
seqfile = 								* sequence data filename
treefile = 								* tree structure file name
outfile = 								* main result file name
noisy = 								* 0,1,2,3,9: how much rubbish on the screen
verbose = 								* 1:detailed output
runmode = 								* 0:user defined tree
seqtype = 								* 1:codons
CodonFreq = 								* 0:equal, 1:F1X4, 2:F3X4, 3:F61
model = 								* 0:one omega ratio for all branches
NSsites = 								* 0:one omega ratio (M0 in Tables 2 and 4)
icode = 								* 0:universal code
fix_kappa = 								* 1:kappa fixed, 0:kappa to be estimated
kappa = 								* initial or fixed kappa
fix_omega = 								* 1:omega fixed, 0:omega to be estimated
omega =  								* initial omega
 ```

In the ```.ctl``` found right up here just few parameters are present and it's possible to customize analyses
even more deeply. 


In this analysis we are going to compare two models:

- m0 where all branches have the same 














We can generate the tree using the command:

```iqtree -s data/example_selection/Rhabdomeric1.fasta```








To run codeml all we need to do is type 'codeml' in the same folder that the codeml.ctl file is in;
all of the options are already in the codeml.ctl file, including all the other inputs required for the analysis. 
Let's type ```Rhabdomeric1.codeml.m0.ctl``` and ```Rhabdomeric1.codeml.m2.ctl```, 
you can finde the ones I used respectively [here](https://github.com/for-giobbe/phy/raw/master/examples/ND2_omega_0.1.ctl) 
and [here](https://github.com/for-giobbe/phy/raw/master/examples/ND2_omega_2.0.ctl).

Let's take a look at the outputs:

* the ```.out``` file which we specified is what we are really interested in.
* ```2NG.t```, ```2NG.dS```, ```2NG.dN``` subs rates, dN & dS distances
* ```4fold.nuc``` 4-fold [degenerate](https://en.wikipedia.org/wiki/Codon_degeneracy) sites
* ```rub```, ```rst```, ```rst1``` and ```lnf``` are some rather misterios intermediate files

The outputs contain many interesting informations, such as:

a summary of codon usage counts:
```
------------------------------------------------------------------------------
Phe F TTT     209 | Ser S TCT     105 | Tyr Y TAT     158 | Cys C TGT      14
      TTC      57 |       TCC      39 |       TAC      45 |       TGC       5
Leu L TTA     579 |       TCA     305 | *** * TAA       0 | Trp W TGA     154
      TTG      14 |       TCG       8 |       TAG       0 |       TGG       7
------------------------------------------------------------------------------
Leu L CTT      46 | Pro P CCT      49 | His H CAT      33 | Arg R CGT      16
      CTC       4 |       CCC      13 |       CAC       9 |       CGC       0
      CTA     102 |       CCA      90 | Gln Q CAA     106 |       CGA      20
      CTG       8 |       CCG       2 |       CAG       5 |       CGG       0
------------------------------------------------------------------------------
Ile I ATT     545 | Thr T ACT     113 | Asn N AAT     414 | Ser S AGT      21
      ATC      92 |       ACC      31 |       AAC      87 |       AGC       3
Met M ATA    1119 |       ACA     222 | Lys K AAA     304 |       AGA     147
      ATG      35 |       ACG       6 |       AAG      10 |       AGG       1
------------------------------------------------------------------------------
Val V GTT      37 | Ala A GCT      42 | Asp D GAT       4 | Gly G GGT      43
      GTC       2 |       GCC       9 |       GAC       2 |       GGC       4
      GTA     118 |       GCA     108 | Glu E GAA     143 |       GGA      97
      GTG       7 |       GCG       1 |       GAG       9 |       GGG      50
------------------------------------------------------------------------------
```

and a summary of base frequencies onthe different codon positions:
```
position  1:    T:0.28185    C:0.08344    A:0.52256    G:0.11214
position  2:    T:0.49336    C:0.18962    A:0.22047    G:0.09655
position  3:    T:0.30674    C:0.06669    A:0.59954    G:0.02704
Average         T:0.36065    C:0.11325    A:0.44752    G:0.07858
```

along with the statistics for each branch in the tree,
```
 branch          t       N       S   dN/dS      dN      dS  N*dN  S*dS

  19..1      2.991   917.2   105.8  0.1000  0.5164  5.1639 473.7 546.1
  19..20     0.528   917.2   105.8  0.1000  0.0913  0.9126  83.7  96.5
  20..2      1.037   917.2   105.8  0.1000  0.1791  1.7914 164.3 189.5
  20..21     0.095   917.2   105.8  0.1000  0.0163  0.1634  15.0  17.3
  21..22     0.125   917.2   105.8  0.1000  0.0216  0.2161  19.8  22.9
  22..23     0.100   917.2   105.8  0.1000  0.0173  0.1727  15.8  18.3
  23..24     0.280   917.2   105.8  0.1000  0.0483  0.4829  44.3  51.1
  24..4      0.512   917.2   105.8  0.1000  0.0885  0.8845  81.1  93.5
  24..25     0.276   917.2   105.8  0.1000  0.0477  0.4769  43.7  50.4
  25..26     0.173   917.2   105.8  0.1000  0.0299  0.2993  27.4  31.6
  26..27     0.102   917.2   105.8  0.1000  0.0175  0.1754  16.1  18.5
  27..28     0.057   917.2   105.8  0.1000  0.0098  0.0984   9.0  10.4
  28..29     0.362   917.2   105.8  0.1000  0.0625  0.6252  57.3  66.1
  29..5      0.139   917.2   105.8  0.1000  0.0240  0.2397  22.0  25.4
  29..6      0.170   917.2   105.8  0.1000  0.0293  0.2930  26.9  31.0
  28..9      0.197   917.2   105.8  0.1000  0.0340  0.3395  31.1  35.9
  27..30     0.027   917.2   105.8  0.1000  0.0046  0.0462   4.2   4.9
  30..31     0.083   917.2   105.8  0.1000  0.0144  0.1435  13.2  15.2
  31..8      0.296   917.2   105.8  0.1000  0.0511  0.5110  46.9  54.0
  31..14     0.351   917.2   105.8  0.1000  0.0606  0.6065  55.6  64.1
  30..32     0.056   917.2   105.8  0.1000  0.0096  0.0963   8.8  10.2
  32..33     0.034   917.2   105.8  0.1000  0.0059  0.0592   5.4   6.3
  33..11     0.235   917.2   105.8  0.1000  0.0406  0.4057  37.2  42.9
  33..12     0.192   917.2   105.8  0.1000  0.0332  0.3316  30.4  35.1
  32..13     0.187   917.2   105.8  0.1000  0.0322  0.3223  29.6  34.1
  26..10     0.148   917.2   105.8  0.1000  0.0255  0.2550  23.4  27.0
  25..7      0.320   917.2   105.8  0.1000  0.0553  0.5531  50.7  58.5
  23..34     0.389   917.2   105.8  0.1000  0.0672  0.6717  61.6  71.0
  34..15     0.636   917.2   105.8  0.1000  0.1098  1.0985 100.8 116.2
  34..16     0.598   917.2   105.8  0.1000  0.1033  1.0330  94.8 109.3
  22..17     1.084   917.2   105.8  0.1000  0.1872  1.8719 171.7 198.0
  21..18     1.176   917.2   105.8  0.1000  0.2031  2.0307 186.3 214.8
  19..3      0.633   917.2   105.8  0.1000  0.1094  1.0936 100.3 115.7

tree length for dN:       2.3465
tree length for dS:      23.4648
```

But we can extract the more important information using respectively:

```
cat ND2_omega_0.1.out | grep lnL
lnL(ntime: 33  np: 34):  -9504.175938      +0.000000
```

and

```
cat ND2_omega_0.1.out | grep lnL
lnL(ntime: 33  np: 34): -11041.325358      +0.000000
```

---

### Task!

 > What if one species was driving the foreground? Try to test a model using m1.

---

### [main](https://github.com/for-giobbe/MP25/tree/main)