### modelling trait evolution on phylogenies: continuous traits


---


### intro: 


Before starting this practical set the appropriate working directory and load the necessary libraries in R:


Libraries:


- [ape](https://emmanuelparadis.github.io/)
- [phytools](http://www.phytools.org/)
- [geiger](https://cran.r-project.org/web/packages/geiger/index.html)


```
setwd("~/Desktop/RPHY/")
library(ape)
library(phytools)
library(geiger)
```


There are no better words than those from [Luke Harmon](https://lukejharmon.github.io/) 
to kickstart this lesson:


>Comparative methods combine biology, mathematics,
>and computer science to learn about a wide variety of topics in evolution using
>phylogenetic trees and other associated data. For example, we can find out which processes must have been
>common, and which rare, across clades in the tree of life; whether evolution
>has proceeded differently in some lineages compared to others; and whether the
>evolutionary potential that we see playing out in real time is sufficient to explain
>the diversity of life on earth, or whether we might need additional processes that
>may come into play only very rarely or over very long timescales,


In this lesson we will perform some ancestral state reconstructions (ASRs) and model-fitting.


---


### inferring a continuous trait ASR: what was the ancestral AT content in Enterobacterales?


The Enterobacterales order of Gammaproteobacteria is a bacteria Order which includes some 
notable species such as E. coli or Y. pestis. They have a huge disparity in lifestyles, 
including several independent shifts to endosymbiosis. In order to infer how their AT content 
maps across their phylogeny, we need to attain a ultrametric tree from a ML one. We will do 
that using the Penalised Likelihood Approach of [chronos](https://www.rdocumentation.org/packages/ape/versions/5.8/topics/chronos) function.
Remember the difference between:


- a tree that has branch lengths that correspond to the amount of change that has happened on each tip and edge
- a tree one in which the distances from the root to every branch tip are equal
- a tree were branch lengths are proportional to time


 To speed up things I arbitrary chose 

- a relaxed clock model to account for heterogeneity among branches.
- seven as lambda that when increased smooths the variation in the rates and tend to a strick clock model (same rate for all branches)


Subsequently we remove possible ploitomies using the multi2di function.


```
tree <- read.tree("trees/tree_enterobacterales.nwk")
tree <- root(tree,c("NZ-CP009610","NZ-CP028926","NZ-CP050969"))
tree <- chronos(tree, model = "relaxed", lambda = 7)
tree <- multi2di(tree)
```


We import the trait states:


```
data <- read.csv("traits/traits_enterobacterales.csv",row.names=1)
AT_content <- setNames(as.numeric(data[,"mean_CDS_at"]),rownames(data))
```


We then perform the ASR of Enterobacterales AT content, which is a continuous trait. Here, to speed up things
we are skipping model selection and assuming a [Brownian Motion](https://en.wikipedia.org/wiki/Brownian_motion), 
yet it is indeed important also when talking about continuous states. Some continuous traits' models include:


- BM - Brownian Motion (BM)
- EB/AC - Early Burst 
- LB/DC - Late Burst 
- OU - Ornstein-Uhlenbeck 
- Trend - BM with a trend 


We can infer and plot the ancestral trait state:


```
fit <- fastAnc(tree, AT_content)
ASR <- contMap(tree,AT_content,plot=FALSE)
plot(ASR, ftype="off",legend=0.5*max(nodeHeights(tree)),
     fsize=0, lwd=3, outline = FALSE)
```


![alt text](https://github.com/for-giobbe/Rphy/blob/main/figures/Figure_5.5.png)


Some times ASRs are not much informative ...


---


### is there a correlation between AT content and the number of genes in Enterobacterales?

Species' traits covary because they share (some) evolutionary history: we expect closely related species 
to be more similar to each other than more distantly related species. 
Treating species as independent in statistical analyses can be misleading! Let see a practical example:
is there a correlation between AT content and the number of genes in Enterobacterales?


Let’s just act as all datapoints (species) are unrelated:

```
fit <- lm(data$mean_CDS_at~data$mean_CDS_ln)
summary(fit)
```


We see that the correlation is significant ... 

```
Call:
lm(formula = data$mean_CDS_at ~ data$mean_CDS_ln)

Residuals:
    Min      1Q  Median      3Q     Max 
-14.947  -9.009  -4.443   7.564  23.845 

Coefficients:
                  Estimate Std. Error t value Pr(>|t|)    
(Intercept)      34.079510   7.370279   4.624 6.61e-06 ***
data$mean_CDS_ln  0.021855   0.007926   2.757  0.00635 ** 
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 11.12 on 208 degrees of freedom
Multiple R-squared:  0.03526,	Adjusted R-squared:  0.03062 
F-statistic: 7.603 on 1 and 208 DF,  p-value: 0.006347
```


But we should instead act as the species considered 
share part of their evolutionary history! The common way to incorporate this
piece of indformation into analyses is the 
Phylogenetic Independent Contrasts (PIC). We will now use Felsenstein’s (1985!) algorithm 
to account for phylogenetic relatedness 
when making interspecific trait comparisons. 


```
mean_CDS_at_PIC<-pic(data$mean_CDS_at,tree)
mean_CDS_ln_PIC<-pic(data$mean_CDS_ln,tree)
fit.pic<-lm(mean_CDS_at_PIC~mean_CDS_ln_PIC)
summary(fit.pic)
```


The p value is different ... 


```
Call:
lm(formula = mean_CDS_at_PIC ~ mean_CDS_ln_PIC)

Residuals:
    Min      1Q  Median      3Q     Max 
-794.89  -39.45    8.09   55.55 1019.47 

Coefficients:
                  Estimate Std. Error t value Pr(>|t|)
(Intercept)     -11.451936  12.775620  -0.896    0.371
mean_CDS_ln_PIC   0.009419   0.008256   1.141    0.255

Residual standard error: 184.7 on 207 degrees of freedom
Multiple R-squared:  0.006249,	Adjusted R-squared:  0.001448 
F-statistic: 1.302 on 1 and 207 DF,  p-value: 0.2552
```


The relationship  retrieved without PIC is a type I error driven by the phylogenetic structure of our data. 
Type II errors are also possible!


---


### a proof of concept on simulated data:

Let's simulate a tree! Everybody should change a bit these numbers

- the random seed (set to 42)
- the number of tips (set to 100)
- the birth rate (set to 1.0)
- the death rate (set to 0.6)
- include only extant - and not extinct - species (set to TRUE)


```
set.seed(42)
tree<-NULL
while(is.null(tree)) 
  tree<-pbtree(n=100,b=1,d=0.6,extant.only=T)
plotTree(tree,ftype="off")
```


Then we can simulate uncorrelated (Brownian) evolution:

```
x <- fastBM(tree)
y <- fastBM(tree)
```


and then plot the resulting trees! 


```
par(mfrow = c(1,2))
obj <- contMap(tree,x,plot=T, ftype="off", lwd=5, outline = FALSE)
obj <- contMap(tree,y,plot=T, ftype="off", lwd=5, outline = FALSE)
```


![alt text](https://github.com/for-giobbe/Rphy/blob/main/figures/Figure_5.6.png)


Now, I want you to perform a linear regression on the two traits, either: 


- entirely disregarding the species phylogeny
- considering species shared evolutionary history 


... we need to take the phylogeny into account ...

---


### further reading: 


[Learning from trees by Luke Harmon](https://lukejharmon.github.io/pcm/) - which is freely available!

[The blog of Liam Revell](http://blog.phytools.org/) - the more fundamental resource in the field.

[the original PIC paper](https://www.jstor.org/stable/2461605?seq=4) - old but gold ..

[dating a tree with ML](https://doi.org/10.1080/10635150802422274) - if your BEASRT2 runs do not converge ehehe ...

[the paper from which I took the stick insects dataset](https://doi.org/10.1093/sysbio/syac038) - I did it :-P


---


[0](https://github.com/for-giobbe/Rphy/blob/main/markdowns/0.md)
[1](https://github.com/for-giobbe/Rphy/blob/main/markdowns/1.md)
[2](https://github.com/for-giobbe/Rphy/blob/main/markdowns/2.md)
[3](https://github.com/for-giobbe/Rphy/blob/main/markdowns/3.md)
[4](https://github.com/for-giobbe/Rphy/blob/main/markdowns/4.md)
[5](https://github.com/for-giobbe/Rphy/blob/main/markdowns/5.md)