### modelling trait evolution on phylogenies: discrete traits


---


### intro: 


Before starting this practical set the appropriate working directory and load the necessary libraries in R:


Libraries:


- [ape](https://emmanuelparadis.github.io/)
- [phytools](http://www.phytools.org/)
- [geiger](https://cran.r-project.org/web/packages/geiger/index.html)


```
setwd("~/Desktop/RPHY/")
library(ape)
library(phytools)
library(geiger)
```


There are no better words than those from [Luke Harmon](https://lukejharmon.github.io/) 
to kickstart this lesson:

### inferring a discrete trait ASR: do Octopodiformes come from the bottom of the ocean?


As always in phylogenetic comparative methods, two fundamental components are required:

 
- a phylogeny - but ultrametric tree is much better
- either discrete or continuous trait(s) 


Here we will leverage a [Octopodiformes](https://en.wikipedia.org/wiki/Octopodiformes) phylogeny and a discrete classification of the maximum depth
at which they have been observed:


- P - pelagic 
- E - euphotic, up to 100 meters in depth
- D1 - disphotic 1, up to 400 meters in depth
- D2 - disphotic 2, up to 400 meters in depth
- N - no-photic, more than 800 meters in depth


Let's start by importing the tree and the data:


```
tree <- read.tree("trees/tree_octopodiformes.nwk")
tree <- force.ultrametric(tree,"extend")
data <- read.table("traits/traits_octopodiformes.csv", sep = ",", row.names = 1, header = T)
data <- setNames(as.factor(data[,"discrete"]),rownames(data))
```


Then, as seen in other instances, we have to find an model appropriate for our data.
Also here, the central component is a rates matrix, not dissimilar from
the Q matrix we have seen for amino acids / nucleotides models.
Let's fit three models to the data, which are:


- a timetree 
- character states


The three models are:

- ER - Equal Rates
- SYM - SYMmetrical
- ARD - All Rates Different


```
ER <- fitDiscrete(tree, data, model="ER")
SYM <- fitDiscrete(tree,data,model="SYM")
ARD <- fitDiscrete(tree,data,model="ARD")
```




Here is a graphical visualisation of the models. 


![alt text](https://github.com/for-giobbe/Rphy/blob/main/figures/Figure_5.1.png)
![alt text](https://github.com/for-giobbe/Rphy/blob/main/figures/Figure_5.2.png)
![alt text](https://github.com/for-giobbe/Rphy/blob/main/figures/Figure_5.3.png)




Then we will asses how good their fit is, using 
[Akaike weights](https://brianomeara.info/aic.html) - the smaller the better!


```
aicc<-setNames(c(ER$opt$aicc,
                 ARD$opt$aicc,
                 SYM$opt$aicc),
               c("Equal Rates",
                 "All Rates Different",
                 "Symmetrical"))
aicw(aicc)
```


We can see that ARD is the better fit!


```
                         fit     delta            w
Equal Rates         123.9865  0.000000 9.893373e-01
All Rates Different 180.3405 56.354017 5.730930e-13
Symmetrical         133.0470  9.060562 1.066273e-02
```


Then we can use the best-fit model to perform the ancestral state reconstruction:


```
fitARD<-ace(data,tree,model="ARD",type="discrete")
```

Then we need to specify a color for each state:


```
levels(data)
cols <- c("red","purple","orange","blue","yellow")

trait_cols <- setNames(c(
                 levels(data)[1],
                 levels(data)[2],
                 levels(data)[3],
                 levels(data)[4],
                 levels(data)[5]),
               c("red",
                 "purple",
                 "orange",
                 "blue",
                 "yellow"))

trait_cols
```


And then plot the ASR on the phylo:


```                 
plot.phylo(tree,lwd=7.5, edge.width = 7.5, label.offset = .2,show.tip.label = T,no.margin = TRUE, 
           edge.color = "lightgray")
nodelabels(node=1:tree$Nnode+Ntip(tree),pie=fitARD$lik.anc,cex=0.42, piecol = cols)
tiplabels(pie=to.matrix(data[tree$tip.label],levels(data)),cex=0.42, piecol = cols)
```


![alt text](https://github.com/for-giobbe/Rphy/blob/main/figures/Figure_5.4.png)


So ... do Octopodiformes come from the photic zone ... or they originated deep down 
in the sea where the light is extremely scarce?

---


### further reading: 


[Learning from trees by Luke Harmon](https://lukejharmon.github.io/pcm/) - which is freely available!

[The blog of Liam Revell](http://blog.phytools.org/) - the more fundamental resource in the field.

[the original PIC paper](https://www.jstor.org/stable/2461605?seq=4) - old but gold ..

[dating a tree with ML](https://doi.org/10.1080/10635150802422274) - if your BEASRT2 runs do not converge ehehe ...

[the paper from which I took the stick insects dataset](https://doi.org/10.1093/sysbio/syac038) - I did it :-P


---


[0](https://github.com/for-giobbe/Rphy/blob/main/markdowns/0.md)
[1](https://github.com/for-giobbe/Rphy/blob/main/markdowns/1.md)
[2](https://github.com/for-giobbe/Rphy/blob/main/markdowns/2.md)
[3](https://github.com/for-giobbe/Rphy/blob/main/markdowns/3.md)
[4](https://github.com/for-giobbe/Rphy/blob/main/markdowns/4.md)
[5](https://github.com/for-giobbe/Rphy/blob/main/markdowns/5.md)