### modelling trait evolution on phylogenies: continuous traits

Before starting this practical set the appropriate working directory and load the necessary libraries in R:

Libraries:

- [ape](https://emmanuelparadis.github.io/)
- [phytools](http://www.phytools.org/)
- [geiger](https://cran.r-project.org/web/packages/geiger/index.html)

```R
setwd("~/Desktop/data/example_PCM")
install.packages('ape')
install.packages('phytools')
install.packages('geiger')
library(ape)
library(phytools)
library(geiger)
```
---

### inferring a continuous trait ASR: what was the ancestral depth 

We import the trait states:


```
data <- read.csv("traits/traits_enterobacterales.csv",row.names=1)
AT_content <- setNames(as.numeric(data[,"mean_CDS_at"]),rownames(data))
```

We then perform the ASR of Enterobacterales AT content, which is a continuous trait. Here, to speed up things
we are skipping model selection and assuming a [Brownian Motion](https://en.wikipedia.org/wiki/Brownian_motion), 
yet it is indeed important also when talking about continuous states. Some continuous traits' models include:

- BM - Brownian Motion (BM)
- EB/AC - Early Burst 
- LB/DC - Late Burst 
- OU - Ornstein-Uhlenbeck 
- Trend - BM with a trend 

We can infer and plot the ancestral trait state:

```
fit <- fastAnc(tree, AT_content)
ASR <- contMap(tree,AT_content,plot=FALSE)
plot(ASR, ftype="off",legend=0.5*max(nodeHeights(tree)),
     fsize=0, lwd=3, outline = FALSE)
```

Some times ASRs are not much informative ...


---


### is there a correlation between depth and eye size?

Species' traits covary because they share (some) evolutionary history: we expect closely related species 
to be more similar to each other than more distantly related species. 
Treating species as independent in statistical analyses can be misleading! Let see a practical example:
is there a correlation between AT content and the number of genes in Enterobacterales?


Let’s just act as all datapoints (species) are unrelated:

```
fit <- lm(data$mean_CDS_at~data$mean_CDS_ln)
summary(fit)
```


We see that the correlation is significant ... 

```
Call:
lm(formula = data$mean_CDS_at ~ data$mean_CDS_ln)

Residuals:
    Min      1Q  Median      3Q     Max 
-14.947  -9.009  -4.443   7.564  23.845 

Coefficients:
                  Estimate Std. Error t value Pr(>|t|)    
(Intercept)      34.079510   7.370279   4.624 6.61e-06 ***
data$mean_CDS_ln  0.021855   0.007926   2.757  0.00635 ** 
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 11.12 on 208 degrees of freedom
Multiple R-squared:  0.03526,	Adjusted R-squared:  0.03062 
F-statistic: 7.603 on 1 and 208 DF,  p-value: 0.006347
```


But we should instead act as the species considered 
share part of their evolutionary history! The common way to incorporate this
piece of indformation into analyses is the 
Phylogenetic Independent Contrasts (PIC). We will now use Felsenstein’s (1985!) algorithm 
to account for phylogenetic relatedness 
when making interspecific trait comparisons. 


```
mean_CDS_at_PIC<-pic(data$mean_CDS_at,tree)
mean_CDS_ln_PIC<-pic(data$mean_CDS_ln,tree)
fit.pic<-lm(mean_CDS_at_PIC~mean_CDS_ln_PIC)
summary(fit.pic)
```


The p value is different ... 


```
Call:
lm(formula = mean_CDS_at_PIC ~ mean_CDS_ln_PIC)

Residuals:
    Min      1Q  Median      3Q     Max 
-794.89  -39.45    8.09   55.55 1019.47 

Coefficients:
                  Estimate Std. Error t value Pr(>|t|)
(Intercept)     -11.451936  12.775620  -0.896    0.371
mean_CDS_ln_PIC   0.009419   0.008256   1.141    0.255

Residual standard error: 184.7 on 207 degrees of freedom
Multiple R-squared:  0.006249,	Adjusted R-squared:  0.001448 
F-statistic: 1.302 on 1 and 207 DF,  p-value: 0.2552
```


The relationship  retrieved without PIC is a type I error driven by the phylogenetic structure of our data. 
Type II errors are also possible!

---

### a proof of concept on simulated data:

Let's simulate a tree! Everybody should change a bit these numbers

- the random seed (set to 42)
- the number of tips (set to 100)
- the birth rate (set to 1.0)
- the death rate (set to 0.6)
- include only extant - and not extinct - species (set to TRUE)


```
set.seed(42)
tree<-NULL
while(is.null(tree)) 
  tree<-pbtree(n=100,b=1,d=0.6,extant.only=T)
plotTree(tree,ftype="off")
```


Then we can simulate uncorrelated (Brownian) evolution:

```
x <- fastBM(tree)
y <- fastBM(tree)
```


and then plot the resulting trees! 


```
par(mfrow = c(1,2))
obj <- contMap(tree,x,plot=T, ftype="off", lwd=5, outline = FALSE)
obj <- contMap(tree,y,plot=T, ftype="off", lwd=5, outline = FALSE)
```


![alt text](https://github.com/for-giobbe/Rphy/blob/main/figures/Figure_5.6.png)


Now, I want you to perform a linear regression on the two traits, either: 


- entirely disregarding the species phylogeny
- considering species shared evolutionary history 


... we need to take the phylogeny into account ...

---

### [main](https://github.com/for-giobbe/MP25/tree/main)
